_posts/2022-01-27-msbuild.md

```diff
@@ -0,0 +1,67 @@
+
+Note: An important detail when using `Directory.Build.props` files is that the file detection stops at the first file found. Therefore, nested `Directory.Build.props` do not work out of the box but you can work around this by specifying `<Import Project="$([MSBuild]::GetPathOfFileAbove($(MSBuildThisFile), $(MSBuildThisFileDirectory)..))" Condition="Exists($([MSBuild]::GetPathOfFileAbove($(MSBuildThisFile), $(MSBuildThisFileDirectory)..)))" />` on the nested file. The analyzer repository makes use of this because it registers a [custom `Directory.Build.props` file](https://github.com/Particular/Particular.Analyzers/blob/master/src/Particular.Analyzers/Directory.Build.props) to **remove** the default dependency on the `Particular.Analyzers` package for itself.
+
+There are also these settings we can briefly look into:
+
+```xml
+<PropertyGroup>
+    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
+    <EnableNETAnalyzers>true</EnableNETAnalyzers>
+    <AnalysisLevel>5.0</AnalysisLevel>
+    <EnforceCodeStyleInBuild>true</EnforceCodeStyleInBuild>
+    <!-- To lock the version of Particular.Analyzers, for example, in a release branch, set this property in Custom.Build.props -->
+    <ParticularAnalyzersVersion Condition="'$(ParticularAnalyzersVersion)' == ''">1.8.0</ParticularAnalyzersVersion>
+    <NServiceBusKey>0024...3b92</NServiceBusKey>
+    <NServiceBusTestsKey>0024...b1c5</NServiceBusTestsKey>
+</PropertyGroup>
+```
+
+This is a bunch of code style related settings we can skim over quickly: `TreatWarningsAsErrors` forces the project build to fail rather than just showing warnings because nobody pays attentions to warnings till shit hits the fan (not just in software development unfortunately). `EnableNETAnalyzers` enables the [built-in Roslyn anlyzers](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/overview) and sets the version of the rules to a specific version using the `AnalysisLevel` property. Note that the Roslyn analyzers are enabled by default anyway, so this is a bit redundant but it's sometimes useful to be explicit. `EnforceCodeStyleInBuild` is another setting related to the Roslyn anlyzers that will allow the analyzers to fail the build (this setting is disabled by default). `ParticularAnalyzersVersion` is similar to `AnalysisLevel` and locks the importet analyzers package version (the version can be overriden by defining a custom version in a `Custom.Build.props` file). The `NServiceBusKey` and `NServiceBusTestsKey` contain the signature of the previously mentioned strong-naming certificates. With those available in being available in msbuild, [`InternalsVisibleTo`](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.internalsvisibletoattribute) can be used more comfortable directly in a `.csproj` file like this (since the assembly exposing internals is strong named, the assemblies allows to access internals must also be strong named):
```

Comment:
```suggestion
Those are a bunch of code-style-related settings we can skim over quickly:

- `TreatWarningsAsErrors` forces the project build to fail rather than just showing warnings because nobody pays attention to warnings till shit hits the fan (not just in software development, unfortunately).
- `EnableNETAnalyzers` enables the [built-in Roslyn anlyzers](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/overview) and sets the version of the rules to a specific version using the `AnalysisLevel` property. Note that the Roslyn analyzers are enabled by default anyway, which is redundant, but it's sometimes helpful to be explicit.
- `EnforceCodeStyleInBuild` is another setting related to the Roslyn analyzers that will allow the analyzers to fail the build (this setting is disabled by default).
- `ParticularAnalyzersVersion` is similar to `AnalysisLevel` and locks the imported analyzers package version (the version can be overridden by defining a custom version in a `Custom.Build.props` file).
- Finally, `NServiceBusKey` and `NServiceBusTestsKey` contain the signature of the previously mentioned strong-naming certificates.

With those available in MSBuild, [`InternalsVisibleTo`](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.internalsvisibletoattribute) can be used directly in a `.csproj` file like this (since the assembly exposing internals is strong named, the assembly allowing to access internals must also be strong-named):
```